### PR TITLE
Julian/backend92 mfa

### LINKED ISSUE #92: Add MFA to login/logout
## Feature Description
Implement a TOTP for login to increase security

blocked by: #91 

## Problem You're Trying to Solve
Users mut be able to log in securely

## Related External Resources
[section 4.4.1](https://docs.google.com/document/d/1DNqexMBf4MpYPCxWsT3plUDAjDSbdJwUzw-UyZX-vDs/edit#heading=h.vbv6udt3y8om)

### PR DESCRIPTION
**Pull request by islandj123**
# Description

Adds backend and frontend functionality for TOTP, including TOTP setup on account creation, and a TOTP code field to be used while logging in.

Closes #92 

## How to Test

Create user endpoint adds the TOTP secret to the database along with the user, and displays a QR-code matching that to the user immediately after creating their account. After the verification email is clicked, then credentials + TOTP code can be used to login.

## Checklist
- [ ] The code includes tests if relevant
- [ ] I have *actually* self-reviewed my changes and done QA

### CODE REVIEW

File: apps/backend/user/src/enableMFA.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
+import { Request, Response } from "express";
+import { IDatabase } from "pg-promise";
+import { User } from "./models/user";
+
+import OTPAuth from "otpauth";
+import * as base32 from "hi-base32";
```
Comment (Scott-Kenning) [inline apps/backend/user/src/enableMFA.ts:6]: just import encode

File: apps/backend/user/src/enableMFA.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  const user = await db.oneOrNone<User>(
+    "SELECT user_id, username, email, password, name, bio, profile_pic_url, verified FROM users WHERE email = $1",
+    [email],
+  );
+
+  if (!user) {
+    return res.status(404).json({ error: "User not found" });
+  }
+
+  if (user.secret) {
```
Comment (Scott-Kenning) [inline apps/backend/user/src/enableMFA.ts:30]: where does user.secret come from

File: apps/backend/user/src/verifyMFA.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  const totp = new OTPAuth.TOTP({
+    label: "Martletplace",
+    algorithm: "SHA1",
+    digits: 6,
+    secret: user.secret!,
+  });
+
+  const invalid = totp.validate({ token });
+
+  if (!invalid && invalid !== null) {
```
Comment (Scott-Kenning) [inline apps/backend/user/src/verifyMFA.ts:37]: check if its === 0, this matches undefined and ""

File: apps/backend/user/src/login.ts [hunk 4]
```diff
@@ -24,8 +27,20 @@ const login = async (req: Request, res: Response, db: IDatabase<object>) => {
...
-    if (!user.verified) {
-      return res.status(401).json({ error: "User is not verified" });
+    // if (!user.verified) {
+    //   return res.status(401).json({ error: "User is not verified" });
+    // }
+
+    // Verify MFA totp token
+    // Only vdo this if user obj has a secret key
+    if (user.totp_secret) {
+      console.log("User has MFA enabled");
```
Comment (MNThomson) [inline apps/backend/user/src/login.ts:37]: User will always have MFA enabled, this is a strict requirement

File: apps/backend/user/src/verifyMFA.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    secret: user.totp_secret!,
+  });
+
+  const isValid = totp.validate({ token: totp_secret });
+
+  if (isValid === 0) {
+    return true;
+  } else {
+    return false;
+  }
```
Comment (MNThomson) [inline apps/backend/user/src/verifyMFA.ts:19]:
```suggestion
  return isValid === 0;
```

File: apps/frontend/src/pages/Login.tsx [hunk 6]
```diff
@@ -11,25 +11,35 @@ const Login = () => {
...
   const [error, setError] = useState<string>("");
   const navigate = useNavigate();
-  const isFormIncomplete = !email || !password;
+  const isFormIncomplete = !email || !password || !totpCode;
   const { setUser } = useUser();
 
   const handleLogin = async (e: FormEvent) => {
     e.preventDefault();
 
+    if (/^\d*$/.test(totpCode) && totpCode.length === 6) {
```
Comment (MNThomson) [inline apps/frontend/src/pages/Login.tsx:23]:
```suggestion
    if (/^\d{6}$/.test(totpCode)) {
```

File: apps/backend/user/package.json [hunk 7]
```diff
@@ -19,6 +19,7 @@
     "@types/express": "^4.17.21",
     "@types/morgan": "^1.9.9",
     "@types/node": "^20.14.4",
+    "@types/qrcode": "^1.5.5",
```
Comment (MNThomson) [inline apps/backend/user/package.json:22]:
```suggestion
```

File: apps/backend/user/package.json [hunk 8]
```diff
@@ -33,7 +34,10 @@
     "bcryptjs": "^2.4.3",
     "cookie-parser": "^1.4.6",
     "express": "^4.19.2",
+    "hi-base32": "^0.5.1",
     "morgan": "^1.10.0",
-    "pg-promise": "^11.8.0"
+    "otpauth": "^9.3.1",
+    "pg-promise": "^11.8.0",
+    "qrcode": "^1.5.3"
```
Comment (MNThomson) [inline apps/backend/user/package.json:41]:
```suggestion
```

**Comment by MNThomson** [comment]
By spec, MFA is enabled for all users (this is not configurable/cannot disable it).
Not sure what the verify mfa endpoint is for, the login endpoint should provide that logic

File: apps/backend/user/src/totpSecretGen.ts [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+  return secret_key;
+};
+
+const generateBase32Secret = () => {
+  const buffer = crypto.randomBytes(15);
+  const base32secret = encode(buffer).toString().replace(/=/g, "");
+  return base32secret.substring(0, 24);
+};
+
+export { totpSecretGen };
```
Comment (MNThomson) [inline apps/backend/user/src/totpSecretGen.ts:18]: Any reason to split this into a different file? I only see this function used in one place

File: apps/backend/user/src/verifyMFA.ts [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+    digits: 6,
+    secret: user.totp_secret!,
+  });
+
+  const isValid = totp.validate({ token: totp_secret });
+
+  return isValid === 0;
+};
+
+export { verifyMFA };
```
Comment (MNThomson) [inline apps/backend/user/src/verifyMFA.ts:18]: Ditto

File: apps/backend/user/src/verifyMFA.ts [hunk 11]
```diff
@@ -0,0 +1,18 @@
+import { User } from "./models/user";
+import { TOTP } from "otpauth";
+
+const verifyMFA = async (user: User, totp_secret: string) => {
```
Comment (MNThomson) [inline apps/backend/user/src/verifyMFA.ts:4]: nit: I prefer `function` declarations. Your call

File: apps/backend/user/src/verifyMFA.ts [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+  // Create TOTP from stored secret
+  const totp = new TOTP({
+    label: "Martletplace",
```
Comment (MNThomson) [inline apps/backend/user/src/verifyMFA.ts:7]:
```suggestion
    label: "martletplace",
```

File: apps/backend/user/src/verifyMFA.ts [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+    algorithm: "SHA1",
+    digits: 6,
+    secret: user.totp_secret!,
```
Comment (MNThomson) [inline apps/backend/user/src/verifyMFA.ts:10]: If this is declared as `string`, why do we need the `!` (non-null assertion)?

File: apps/backend/user/src/verifyMFA.ts [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+  });
+
+  const isValid = totp.validate({ token: totp_secret });
```
Comment (MNThomson) [inline apps/backend/user/src/verifyMFA.ts:13]: Should this var be named `isValid` then if its not a bool or 0/1? What's the possible values for it?

File: apps/backend/user/tests/login.test.ts [hunk 15]
```diff
@@ -3,19 +3,31 @@ import { login } from '../src/login';
...
 vi.mock('../../lib/src/auth', () => ({
   create_token: vi.fn(),
 }));
 
-describe('Login Endpoint', () => {
+describe('Login Endpoint', () => { 
   it('should login a user successfully', async () => {
+    // Generate totp from test user secret
+    const totp = new TOTP({
+      label: "Martletplace",
```
Comment (MNThomson) [inline apps/backend/user/tests/login.test.ts:18]:
```suggestion
      label: "martletplace",
```

File: apps/frontend/src/pages/CreateAccount.tsx [hunk 16]
```diff
@@ -6,12 +6,15 @@ import {
...
   FormHelperText,
   Link,
+  Backdrop,
 } from "@mui/material";
 import martletPlaceLogo from "../images/martletplace-logo.png";
 import { colors } from "../styles/colors";
 import { useStyles } from "../styles/pageStyles";
 import { useNavigate } from "react-router-dom";
 import axios from "axios";
+import * as OTPAuth from "otpauth";
```
Comment (MNThomson) [inline apps/frontend/src/pages/CreateAccount.tsx:16]: Lets import the specific things we need

File: apps/frontend/src/pages/CreateAccount.tsx [hunk 17]
```diff
@@ -24,6 +27,37 @@ const CreateAccount = () => {
   const [usernameError, setUsernameError] = useState("");
   const [passwordError, setPasswordError] = useState("");
 
+  const [openBackDrop, setOpenBackDrop] = useState(false);
+  const [tokenCode, setTokenCode] = useState("");
+  const [qrCodeUrl, setQrCodeUrl] = useState("");
+
+  const generateCode = (totp_secret: string, email: string) => {
+    const totp = new OTPAuth.TOTP({
+      issuer: "MartletPlace",
```
Comment (MNThomson) [inline apps/frontend/src/pages/CreateAccount.tsx:36]:
```suggestion
      issuer: "martletplace",
```

File: apps/backend/user/src/verifyMFA.ts (hunk 12 above)
Comment (islandj123) [reply apps/backend/user/src/verifyMFA.ts:7]: I think this is only important for the CreateAccount.tsx appearance of this, but that exact string is displayed as the label in auth apps so it makes sense to keep the capitalization of "MartletPlace". It should still be made consistent however.

File: apps/backend/user/src/verifyMFA.ts (hunk 14 above)
Comment (islandj123) [reply apps/backend/user/src/verifyMFA.ts:13]: That validate method returns the token delta or possibly null if very invalid, so I've renamed the const variable to better reflect that.

**Review by Scott-Kenning — APPROVED** [review]

**Review by CaelumD25 — APPROVED** [review]
Looks alright to me in respect to frontend!

**Review by MNThomson — APPROVED** [review]

**Pull request merged**